File: python/aethernet/market/pov_token.py
# SPDX-License-Identifier: MIT
#
# Proof-of-Vicinity token model and canonical signable-body codec. Python port of
# AetherNet.Market.Models.PoVToken / PoVTransportType / PoVScore and AetherNet.Market.PoVTokenCodec.
#
# The canonical body that BOTH the witness and the subject sign with their real Ed25519 identity keys
# must stay byte-identical across every language implementation so a token signed by one node verifies
# on any other:
#
#   SubjectLen(4 LE i32) || Subject(UTF-8) || TimestampTicks(8 LE i64) || Transport(1 byte)
#
# timestamp_ticks is .NET DateTime.Ticks (100ns intervals since 0001-01-01).

from __future__ import annotations

import json
import struct
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from enum import IntEnum
from typing import Optional

# Number of .NET DateTime ticks (100ns) per second.
_TICKS_PER_SECOND = 10_000_000

# The .NET DateTime.Ticks value at the Unix epoch (1970-01-01T00:00:00Z) — i.e. ticks between
# 0001-01-01 and 1970-01-01. Used to convert between .NET ticks and Python datetime.
_UNIX_EPOCH_TICKS = 621_355_968_000_000_000


class PoVTransportType(IntEnum):
    """The transport used for a co-presence Proof-of-Vicinity exchange. Only short-range transports are
    valid (prevents remote minting)."""

    # Bluetooth Low Energy (short range — prevents remote forgery).
    Ble = 0
    # Near-Field Communication (requires physical proximity).
    Nfc = 1
    # Huawei NearLink (short range, similar to BLE).
    NearLink = 2

    def is_short_range(self) -> bool:
        """Whether the transport is a valid short-range PoV channel."""
        return self in (PoVTransportType.Ble, PoVTransportType.Nfc, PoVTransportType.NearLink)

    @property
    def wire_name(self) -> str:
        """The lowercase wire name of the transport (``ble`` / ``nfc`` / ``nearlink``)."""
        return {
            PoVTransportType.Ble: "ble",
            PoVTransportType.Nfc: "nfc",
            PoVTransportType.NearLink: "nearlink",
        }[self]


def build_signable_token_data(
    subject_uhid: str, timestamp_ticks: int, transport: PoVTransportType
) -> bytes:
    """Build the canonical signable bytes for a PoV token body. The same layout is signed by the witness
    (on issue) and counter-signed by the subject (on accept)::

        SubjectLen(4 LE i32) || Subject(UTF-8) || TimestampTicks(8 LE i64) || Transport(1 byte)
    """
    subject_bytes = subject_uhid.encode("utf-8")
    return (
        struct.pack("<i", len(subject_bytes))
        + subject_bytes
        + struct.pack("<q", timestamp_ticks)
        + bytes([int(transport) & 0xFF])
    )


@dataclass
class PoVToken:
    """A Proof-of-Vicinity token issued by one node (the witness) to another (the subject) during a
    physical co-presence event.

    Both parties must countersign — this prevents unilateral forgery. The token is transmitted over a
    short-range transport (BLE/NFC/NearLink only) to prevent remote minting. The JSON wire form is
    snake_case, matching the C# serializer.
    """

    # UHID of the node issuing the voucher.
    witness_uhid: str = ""

    # UHID of the node being vouched for.
    subject_uhid: str = ""

    # The co-presence event time as .NET DateTime.Ticks (100ns since 0001-01-01). Stored as ticks (not a
    # Python datetime) so the signed canonical body is byte-identical to C#.
    timestamp_ticks: int = 0

    # The transport channel used (must be short-range).
    transport_used: PoVTransportType = PoVTransportType.Ble

    # Ed25519 signature by the witness over the canonical body.
    witness_signature: Optional[bytes] = None

    # Ed25519 countersignature by the subject — required for token validity.
    subject_signature: Optional[bytes] = None

    def signable_data(self) -> bytes:
        """The canonical signable bytes for this token."""
        return build_signable_token_data(
            self.subject_uhid, self.timestamp_ticks, self.transport_used
        )

    def to_json(self) -> bytes:
        """Serialise the token to its snake_case UTF-8 JSON wire form. Signatures are hex (or omitted
        when absent)."""
        obj: dict = {
            "witness_uhid": self.witness_uhid,
            "subject_uhid": self.subject_uhid,
            "timestamp_ticks": self.timestamp_ticks,
            "transport_used": int(self.transport_used),
        }
        if self.witness_signature is not None:
            obj["witness_signature"] = self.witness_signature.hex()
        if self.subject_signature is not None:
            obj["subject_signature"] = self.subject_signature.hex()
        return json.dumps(obj).encode("utf-8")

    @classmethod
    def from_json(cls, data: bytes) -> "PoVToken":
        """Deserialise a snake_case UTF-8 JSON PoV token."""
        obj = json.loads(data.decode("utf-8") if isinstance(data, (bytes, bytearray)) else data)
        wsig = obj.get("witness_signature")
        ssig = obj.get("subject_signature")
        return cls(
            witness_uhid=obj.get("witness_uhid", ""),
            subject_uhid=obj.get("subject_uhid", ""),
            timestamp_ticks=int(obj.get("timestamp_ticks", 0)),
            transport_used=PoVTransportType(int(obj.get("transport_used", 0))),
            witness_signature=bytes.fromhex(wsig) if wsig else None,
            subject_signature=bytes.fromhex(ssig) if ssig else None,
        )


def ticks_to_datetime(ticks: int) -> datetime:
    """Convert a .NET DateTime.Ticks value to a timezone-aware UTC ``datetime``. Provided for hosts that
    want a Python datetime; the canonical body always uses the raw ticks."""
    unix_ticks = ticks - _UNIX_EPOCH_TICKS
    micros = unix_ticks // 10  # 100ns ticks -> microseconds
    return datetime(1970, 1, 1, tzinfo=timezone.utc) + timedelta(microseconds=micros)


def datetime_to_ticks(dt: datetime) -> int:
    """Convert a ``datetime`` to a .NET DateTime.Ticks value. Naive datetimes are assumed to be UTC."""
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    delta = dt.astimezone(timezone.utc) - datetime(1970, 1, 1, tzinfo=timezone.utc)
    # delta.microseconds resolution is 1us = 10 ticks; sub-microsecond .NET precision is not
    # representable in a Python datetime, so this round-trips at microsecond (10-tick) resolution.
    total_micros = (delta.days * 86_400 + delta.seconds) * 1_000_000 + delta.microseconds
    return total_micros * 10 + _UNIX_EPOCH_TICKS


@dataclass
class PoVScore:
    """The Proof-of-Vicinity trust score for a node — a purely local anti-Sybil routing/identity signal
    that attaches NO value semantics."""

    # UHID of the scored node.
    uhid: str = ""
    # Number of distinct witnesses who have issued PoV tokens to this node.
    unique_witnesses: int = 0
    # Weighted score (0.0–1.0).
    weighted_score: float = 0.0
    # Time of the most recent score update.
    last_updated: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

File: kotlin/src/main/kotlin/aethernet/market/PoVToken.kt
// SPDX-License-Identifier: MIT
//
// Proof-of-Vicinity token model and canonical signable-body codec. Kotlin port of
// AetherNet.Market.Models.PoVToken / PoVTransportType / PoVScore and AetherNet.Market.PoVTokenCodec,
// byte-identical to the C# reference and the Go port.
//
// The canonical body that BOTH the witness and the subject sign with their real Ed25519 identity keys
// must stay byte-identical across every language implementation so a token signed by one node verifies
// on any other:
//
//   SubjectLen(4 LE i32) || Subject(UTF-8) || TimestampTicks(8 LE i64) || Transport(1 byte)
//
// timestamp_ticks is .NET DateTime.Ticks (100ns intervals since 0001-01-01).

package aethernet.market

import aethernet.incentive.appendJsonString
import org.json.JSONObject
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * The transport used for a co-presence Proof-of-Vicinity exchange. Only short-range transports are
 * valid (prevents remote minting). Enum wire bytes: ble = 0, nfc = 1, nearlink = 2.
 */
enum class PoVTransportType(val value: Byte) {
    /** Bluetooth Low Energy (short range — prevents remote forgery). */
    Ble(0),

    /** Near-Field Communication (requires physical proximity). */
    Nfc(1),

    /** Huawei NearLink (short range, similar to BLE). */
    NearLink(2);

    /** Whether the transport is a valid short-range PoV channel. */
    fun isShortRange(): Boolean = when (this) {
        Ble, Nfc, NearLink -> true
    }

    /** The lowercase wire name of the transport. */
    fun wireName(): String = when (this) {
        Ble -> "ble"
        Nfc -> "nfc"
        NearLink -> "nearlink"
    }

    companion object {
        /** Resolves a transport from its wire byte, or null if unknown (not short-range). */
        fun fromValue(value: Byte): PoVTransportType? = entries.find { it.value == value }
    }
}

/**
 * A Proof-of-Vicinity token issued by one node (the witness) to another (the subject) during a
 * physical co-presence event. Both parties must countersign — this prevents unilateral forgery. The
 * token is transmitted over a short-range transport (BLE/NFC/NearLink only) to prevent remote minting.
 *
 * Wire format: hand-rolled snake_case JSON ([buildString] encode, [org.json.JSONObject] decode) rather
 * than kotlinx.serialization, so this type compiles under AOSP Soong's plain kotlinc. Canonical key
 * order: `witness_uhid`, `subject_uhid`, `timestamp_ticks`, `transport_used`, `witness_signature`,
 * `subject_signature`.
 */
data class PoVToken(
    /** UHID of the node issuing the voucher. Wire key: `witness_uhid`. */
    val witnessUhid: String = "",

    /** UHID of the node being vouched for. Wire key: `subject_uhid`. */
    val subjectUhid: String = "",

    /**
     * Co-presence event time as .NET `DateTime.Ticks` (100ns since 0001-01-01). Stored as ticks
     * (not a Java time type) so the signed canonical body is byte-identical to C#. Wire key:
     * `timestamp_ticks`.
     */
    val timestampTicks: Long = 0L,

    /** Transport channel used (must be short-range). Wire key: `transport_used`. */
    val transportUsed: PoVTransportType = PoVTransportType.Ble,

    /** Ed25519 signature by the witness over the canonical body. Wire key: `witness_signature`. */
    val witnessSignature: ByteArray? = null,

    /**
     * Ed25519 countersignature by the subject — required for token validity. Wire key:
     * `subject_signature`.
     */
    val subjectSignature: ByteArray? = null,
) {
    /** The canonical signable bytes for this token. */
    fun signableData(): ByteArray = buildSignableTokenData(subjectUhid, timestampTicks, transportUsed)

    /** Canonical snake_case wire JSON. Fixed key order — see class doc. */
    fun toJson(): String = buildString {
        append("{\"witness_uhid\":"); appendJsonString(witnessUhid)
        append(",\"subject_uhid\":"); appendJsonString(subjectUhid)
        append(",\"timestamp_ticks\":").append(timestampTicks)
        append(",\"transport_used\":").append(transportUsed.value.toInt())
        append(",\"witness_signature\":")
        if (witnessSignature == null) append("null") else appendJsonString(base64(witnessSignature))
        append(",\"subject_signature\":")
        if (subjectSignature == null) append("null") else appendJsonString(base64(subjectSignature))
        append('}')
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PoVToken) return false
        if (witnessUhid != other.witnessUhid) return false
        if (subjectUhid != other.subjectUhid) return false
        if (timestampTicks != other.timestampTicks) return false
        if (transportUsed != other.transportUsed) return false
        if (!sigEquals(witnessSignature, other.witnessSignature)) return false
        if (!sigEquals(subjectSignature, other.subjectSignature)) return false
        return true
    }

    override fun hashCode(): Int {
        var result = witnessUhid.hashCode()
        result = 31 * result + subjectUhid.hashCode()
        result = 31 * result + timestampTicks.hashCode()
        result = 31 * result + transportUsed.hashCode()
        result = 31 * result + (witnessSignature?.contentHashCode() ?: 0)
        result = 31 * result + (subjectSignature?.contentHashCode() ?: 0)
        return result
    }

    companion object {
        /** Number of .NET DateTime ticks (100ns) per second. */
        const val TICKS_PER_SECOND = 10_000_000L

        /**
         * The .NET `DateTime.Ticks` value at the Unix epoch (1970-01-01T00:00:00Z), i.e. ticks
         * between 0001-01-01 and 1970-01-01. Used to convert between .NET ticks and Unix time.
         */
        const val UNIX_EPOCH_TICKS = 621_355_968_000_000_000L

        /**
         * Builds the canonical signable bytes for a PoV token body. The same layout is signed by the
         * witness (on issue) and counter-signed by the subject (on accept).
         * ```
         *   SubjectLen(4 LE i32) || Subject(UTF-8) || TimestampTicks(8 LE i64) || Transport(1 byte)
         * ```
         */
        fun buildSignableTokenData(
            subjectUhid: String,
            timestampTicks: Long,
            transport: PoVTransportType,
        ): ByteArray {
            val subjectBytes = subjectUhid.toByteArray(Charsets.UTF_8)
            val buffer = ByteBuffer.allocate(4 + subjectBytes.size + 8 + 1).order(ByteOrder.LITTLE_ENDIAN)
            buffer.putInt(subjectBytes.size)
            buffer.put(subjectBytes)
            buffer.putLong(timestampTicks)
            buffer.put(transport.value)
            return buffer.array()
        }

        /** Parse from canonical JSON. Returns null on malformed / missing-field input. */
        fun fromJson(json: String): PoVToken? = try {
            val o = JSONObject(json)
            val transport = PoVTransportType.fromValue(o.getInt("transport_used").toByte())
                ?: return null
            val wSig = if (!o.has("witness_signature") || o.isNull("witness_signature")) null
            else unbase64(o.getString("witness_signature"))
            val sSig = if (!o.has("subject_signature") || o.isNull("subject_signature")) null
            else unbase64(o.getString("subject_signature"))
            PoVToken(
                witnessUhid = o.getString("witness_uhid"),
                subjectUhid = o.getString("subject_uhid"),
                timestampTicks = o.getLong("timestamp_ticks"),
                transportUsed = transport,
                witnessSignature = wSig,
                subjectSignature = sSig,
            )
        } catch (_: Exception) {
            null
        }

        /**
         * Converts a .NET `DateTime.Ticks` value to Unix milliseconds (UTC). Provided for hosts that
         * want wall-clock time; the canonical body always uses the raw ticks.
         */
        fun ticksToUnixMillis(ticks: Long): Long = (ticks - UNIX_EPOCH_TICKS) / 10_000L

        /** Converts Unix milliseconds to a .NET `DateTime.Ticks` value. */
        fun unixMillisToTicks(unixMillis: Long): Long = unixMillis * 10_000L + UNIX_EPOCH_TICKS

        private fun sigEquals(a: ByteArray?, b: ByteArray?): Boolean =
            if (a == null) b == null else b != null && a.contentEquals(b)

        private fun base64(b: ByteArray): String = java.util.Base64.getEncoder().encodeToString(b)
        private fun unbase64(s: String): ByteArray = java.util.Base64.getDecoder().decode(s)
    }
}

/**
 * The Proof-of-Vicinity trust score for a node — a purely local anti-Sybil routing/identity signal
 * that attaches NO value semantics.
 */
data class PoVScore(
    /** UHID of the scored node. */
    val uhid: String,
    /** Number of distinct witnesses who have issued PoV tokens to this node. */
    val uniqueWitnesses: Int,
    /** Weighted score (0.0–1.0). */
    val weightedScore: Double,
    /** Time of the most recent score update, in Unix milliseconds. */
    val lastUpdatedUnixMs: Long,
)
